cupy/cuda/nccl.pyx

```diff
@@ -128,14 +134,90 @@ def get_unique_id():
     return ret
 
 
+cpdef groupStart():
+    """Start a group of NCCL calls. Must be paired with :func:`groupEnd()`.
+
+    .. note::
+        This method is only useful when the ``NcclCommunicator`` instances are
+        created via :meth:`~.NcclCommunicator.initAll`. A typical usage pattern
+        is like this:
+
+        .. code-block:: python
+
+            comms = cupy.cuda.nccl.NcclCommunicator.initAll(n, dev_list)
+            # ... do some preparation work
+            cupy.cuda.nccl.groupStart()
+            for rank, comm in enumerate(comms):
+                # ... make some collective calls ...
+            cupy.cuda.nccl.groupEnd()
+
+    .. seealso:: `ncclGroupStart`_
+
+    .. _ncclGroupStart:
+        https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/api/group.html#ncclgroupstart
+    """  # noqa
+    with nogil:
+        status = ncclGroupStart()
+    check_status(status)
+
+
+cpdef groupEnd():
+    """End a group of NCCL calls. Must be paired with :func:`groupStart()`.
+
+    .. note::
+        This method is only useful when the ``NcclCommunicator`` instances are
+        created via :meth:`~.NcclCommunicator.initAll`. A typical usage pattern
+        is like this:
+
+        .. code-block:: python
+
+            comms = cupy.cuda.nccl.NcclCommunicator.initAll(n, dev_list)
+            # ... do some preparation work
+            cupy.cuda.nccl.groupStart()
+            for rank, comm in enumerate(comms):
+                # ... make some collective calls ...
+            cupy.cuda.nccl.groupEnd()
+
+    .. seealso:: `ncclGroupEnd`_
+
+    .. _ncclGroupEnd:
+        https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/api/group.html#ncclgroupend
+    """  # noqa
+    with nogil:
+        status = ncclGroupEnd()
+    check_status(status)
+
+
 cdef class NcclCommunicator:
+    """ Initialize an NCCL communicator for one device controlled by one
+    process.
+
+    Args:
+        ndev (int): Total number of GPUs to be used.
+        commId (tuple): The unique ID returned by :func:`get_unique_id`.
+        rank (int): The rank of the GPU managed by the current process.
+
+    Returns:
+        NcclCommunicator: An ``NcclCommunicator`` instance.
+
+    .. note::
+        This method is for creating an NCCL communicator in a multi-process
+        environment, typically managed by MPI or ``multiprocessing``.
+
+    .. seealso:: `ncclCommInitRank`_
+
+    .. _ncclCommInitRank:
+        https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/api/comms.html#ncclcomminitrank
+    """  # noqa
 
     cdef:
         ncclComm_t _comm
 
+    def __cinit__(self):
```

Comment:
I think the description in this section could have been clearer. By that they really mean "any C attributes should be initialized in `__cinit__()`". Similar poor description can be found for `__init__()`:
> By the time `__init__()` is called, the object is a fully valid Python object and all operations are safe.

But apparently without proper initialization it's impossible for all operations to be safe 😛 Also, we could not rule out the possibility that without calling `__cinit__()` the C attributes are not (zero) initialized 😄 

In this case, while `_comm` is implicitly zero-initialized when `__cinit__()` is removed, this behavior should not be taken as guaranteed: This can be checked by removing `__cinit__()` and examining the generated `nccl.cpp` file. `__pyx_v_self->_comm` is never properly initialized. (NCCL might have done that under the hood for us, but it shouldn't be relied on either.) This is not good if the code path along `__init__()` is taken.

Finally, it seems to be a common practice to call `__cinit__()` for `cdef`'d classes, see https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#fast-instantiation and https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#instantiation-from-existing-c-c-pointers

Reply:
Actually, I suppose, NcclCommunicator is allocated by `PyType_GenericAlloc` with the value of its type object's `tp_basicsize`:

In nccl.cpp:
```CPP
static PyTypeObject __pyx_type_4cupy_4cuda_4nccl_NcclCommunicator = {
  ...
  sizeof(struct __pyx_obj_4cupy_4cuda_4nccl_NcclCommunicator), /*tp_basicsize*/
  ...
}
```
```CPP
struct __pyx_obj_4cupy_4cuda_4nccl_NcclCommunicator {
  PyObject_HEAD
  struct __pyx_vtabstruct_4cupy_4cuda_4nccl_NcclCommunicator *__pyx_vtab;
  ncclComm_t _comm;
};
```

, and `PyType_GenericAlloc` allocates a new instance and clears its contents to `\0` with the size here:
https://github.com/python/cpython/blob/v3.7.4/Objects/typeobject.c#L999

But I also think your explanation makes sense and explicitly initializing `_comm` with zero has no harm as well as being safer, so let's leave the `__cinit__()` as it is. Thanks.

Reply:
Thanks for pointing me to Cython's internal (`PyType_GenericAlloc`). It's always cool to learn more! Mark this as resolved.